### PR TITLE
feat: Add option to conditionally disable site instance patches

### PR DESCRIPTION
**Pull request by MarshallOfSound**
#### Description of Change
This adds a new `app.disableElectronSiteInstanceOverrides` that allows us to dynamically (at runtime) disable the site instance override patches.  The intention is to:
* have this PR land in Electron 6
* have the first deprecation warning to land in Electron 7 for non-context aware native modules
* have a deprecation warning land in Electron 8 for the default value of `app.disableElectronSiteInstanceOverrides` to switch
* switch the default value of `app.disableElectronSiteInstanceOverrides` in Electron 9
* Deprecate the ability to change `app.disableElectronSiteInstanceOverrides` in Electron 10
* Remove the ability to change `app.disableElectronSiteInstanceOverrides` in Electron 11

Yes you're reading that right, 5 major versions 👍 (60~ish weeks)

For more info see this [Discussion Issue](https://github.com/electron/electron/issues/18397)

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Still TODO

- [x] Find a way to prevent loading native modules that are not context aware / NAPI in the renderer process if this new mode is enabled.  **Required for this PR**
- [x] Add a `console.warn` for loading native modules in the renderer process that are not context-aware or NAPI - This warning will ship in Electron 7

#### Release Notes

Notes: Added new `app.disableElectronSiteInstanceOverrides` property that allows apps to disable the site instance overrides Electron has patched into Chromium.  This can be used to prevent the automatic renderer process restarting that Electron currently does.

### CODE REVIEW

File: patches/common/chromium/frame_host_manager.patch [hunk 1]
```diff
@@ -4,94 +4,104 @@ Date: Wed, 14 Nov 2018 20:38:46 +0530
...
-+      delegate_->GetControllerForRenderManager().GetBrowserContext();
-+  // If the navigation can swap SiteInstances, compute the SiteInstance it
-+  // should use.
-+  // TODO(clamy): We should also consider as a candidate SiteInstance the
-+  // speculative SiteInstance that was computed on redirects.
-+  scoped_refptr<SiteInstance> candidate_site_instance =
-+      speculative_render_frame_host_
-+          ? speculative_render_frame_host_->GetSiteInstance()
-+          : nullptr;
++  BrowserContext* browser_context;
```
Comment (miniak) [inline patches/common/chromium/frame_host_manager.patch]: `BrowserContext* browser_context = nullptr:` ?

File: DEPS [hunk 2]
```diff
@@ -12,7 +12,7 @@ vars = {
   'chromium_version':
     '84c40395c741fa24ccbd9fc2c5828e2e97472952',
   'node_version':
-    'a86a4a160dc520c61a602c949a32a1bc4c0fc633',
+    'b213b26f0fb642b95561c0f10039886607036d5a',
```
Comment (nornagon) [inline DEPS]: link to electron/nodejs pr?

Comment (MarshallOfSound) [reply DEPS]: Tis just this commit --> https://github.com/electron/node/commit/b213b26f0fb642b95561c0f10039886607036d5a

File: atom/browser/api/atom_api_app.cc [hunk 3]
```diff
@@ -1286,6 +1286,13 @@ std::string App::GetUserAgentFallback() {
   return AtomBrowserClient::Get()->GetUserAgent();
 }
 
+void App::SetDisableElectronSiteInstanceOverrides(bool should_disable) {
```
Comment (deepak1556) [inline atom/browser/api/atom_api_app.cc]: Better naming `App::DisableCustomSiteInstance` ?

File: atom/browser/api/atom_api_app.cc [hunk 4]
```diff
@@ -1286,6 +1286,13 @@ std::string App::GetUserAgentFallback() {
...
+  AtomBrowserClient::Get()->SetDisableProcessRestartTricks(should_disable);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_app.cc]: Can use the same name `AtomBrowserClient::DisableCustomSiteInstance`

File: atom/browser/api/atom_api_app.cc [hunk 5]
```diff
@@ -1286,6 +1286,13 @@ std::string App::GetUserAgentFallback() {
...
+}
+bool App::GetDisableElectronSiteInstanceOverrides() {
```
Comment (deepak1556) [inline atom/browser/api/atom_api_app.cc]: `App::CanUseCustomSiteInstance`

File: atom/browser/api/atom_api_app.cc [hunk 6]
```diff
@@ -1286,6 +1286,13 @@ std::string App::GetUserAgentFallback() {
...
+  return AtomBrowserClient::Get()->ShouldDisableElectronProcessRestartTricks();
```
Comment (deepak1556) [inline atom/browser/api/atom_api_app.cc]: `AtomBrowserClient::CanUseCustomSiteInstance`

**Comment by 396784797** [comment]
app.allowRendererProcessReuse = true ；pall on electron9.0.0 nodejs 12.17.0

**Pull request merged**